packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -1,8 +1,72 @@
 pragma solidity ^0.6.0;
 pragma experimental ABIEncoderV2;
 
+import "../interfaces/FundingRateStoreInterface.sol";
+import "../../../common/implementation/Testable.sol";
 
-// TODO:
-contract FundingRateStore {
 
+contract FundingRateStore is FundingRateStoreInterface, Testable {
+    struct Proposal {
+        FixedPoint.Signed rate;
+        uint256 time;
+        address proposer;
+    }
+
+    struct FundingRateRecord {
+        FixedPoint.Signed rate;
+        Proposal proposal;
+    }
+
+    enum ProposalState { None, Pending, Expired }
+
+    mapping(bytes32 => FundingRateRecord) public fundingRateRecords;
+
+    uint256 public proposalLiveness;
+
+    constructor(uint256 _proposalLiveness, address _timerAddress) public Testable(_timerAddress) {
+        require(_proposalLiveness > 0, "Proposal liveness is 0");
+        proposalLiveness = _proposalLiveness;
+    }
+
+    function getFundingRateForIdentifier(bytes32 identifier) external override view returns (FixedPoint.Signed memory) {
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(identifier);
+
+        if (_getProposalState(fundingRateRecord.proposal) == ProposalState.Expired) {
+            return fundingRateRecord.proposal.rate;
+        } else {
+            return fundingRateRecord.rate;
+        }
+    }
+
+    function propose(bytes32 identifier, FixedPoint.Signed memory rate) external {
+        // TODO: ACLS on identifiers that can be proposed.
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(identifier);
+
+        ProposalState proposalState = _getProposalState(fundingRateRecord.proposal);
+
+        // TODO: bond logic.
+
+        require(proposalState != ProposalState.Pending, "Existing proposal still pending.");
+
+        if (proposalState == ProposalState.Expired) {
```

Comment:
So, to "set" an expired funding rate as the current one, `propose()` must be called after the previously proposed funding rate is expired? I agree with this thinking, just want to be clear I'm understanding.

Reply:
The way I'm thinking about it (and this may be a little confusing) is that the new funding rate goes into effect immediately once the proposal passes liveness, but it doesn't get moved in storage until this proposal is called, since it doesn't need to be.

Reply:
Yeah that's my understanding of your code, which I think makes sense.

Reply:
SG!